### PR TITLE
Customisable additional directory

### PR DESCRIPTION
**Pull request by maciekberry**
Hello,

I restarted to work on the better usability of the Media Library. Please find below my first pull request. It enables a new variable in the VirtualFolders.conf file: addToMediaLibrary. By default it is true and the behavior of UMS is not changed. If the user sets it to false (on per directory basis), the directory is added to UMS but the contents will now show up in the Media Library. It is extremely useful if you are maniac about the music library pureness and you have some folders with the music that is not properly tagged.

If you accept the pull request, I can also add some documentation (just tell me where).

As usual, if the PR is not clean enough, please do not shoot, I am a github rookie. :(

### CODE REVIEW

File: src/main/java/net/pms/dlna/DLNAResource.java [hunk 1]
```diff
@@ -5155,5 +5155,8 @@ private void setMetadataFromFileName(File file) {
 				OpenSubtitle.backgroundLookupAndAdd(file, media);
 			}
 		}
+
```
Comment (valib) [inline src/main/java/net/pms/dlna/DLNAResource.java]: @maciekberry here is missing `}`

**Comment by maciekberry** [comment]
I should be punished for requesting a merge that does not compile :(

I see it passed all checks now. Sorry again.

**Comment by valib** [comment]
@maciekberry I tried to understand the code but for me is not clear where is the GUI for users to set whether the directory should be added the Media Library or not.

**Comment by maciekberry** [comment]
Hello @valib . As far as I understand, there is no GUI for declaring the Virtual Folder. This happens manualy in the "VirtualFolders.conf" file (or directly in the UMS.conf in "virtual_folders" variable). This is NOT for main folders (the ones you can declare from the GUI) which are always included in Media Library.

**Comment by SubJunk** [comment]
@maciekberry thanks for this. It looks good. Can you please also document it in UMS.conf and put an example of it in VirtualFolders.conf so there is more visibility over this new option?

**Comment by SubJunk** [comment]
@maciekberry is this still active?

**Comment by maciekberry** [comment]
Yes, I am sorry - was busy elsewhere. Will commit soon.

**Comment by maciekberry** [comment]
Hello,
Documentation done. As written, the feature is enabled only if you use VirtualFolders.conf file. The syntax of the UMS.conf variable is already complicated enough, I have decided not to complicate it even more.

**Comment by SubJunk** [comment]
@maciekberry thanks for this, the change looks good, however something happened with Git, maybe when you synchronised with our `master`, so now if we merge it will roll back some of our code. 

Maybe you could make a new branch off the latest `master` and then `cherry-pick` your commits `b8a4ac2`, `2ab5722` and `3cbda6d` into it?

**Comment by SubJunk** [comment]
@maciekberry if you want I can try to do that for you

**Comment by maciekberry** [comment]
I am sorry - I missed your comment from 14 days ago :((((

To answer your question: yes, please, can you do it? - I still have problems with git. I promise I will follow a training soon. It's really shameful.

**Comment by maciekberry** [comment]
Please @SubJunk what is the status of that? Can I do anything to help?

**Comment by Nadahar** [comment]
@maciekberry Git is both difficult and very easy at the same time. There are lots of advanced stuff you can do that most developers don't have a clue how to use. The basic stuff is what you normally need though, and that's very easy once you get over a certain threshold. You have to get the fundamental idea of how it works, and until you do, nothing with Git will make much sense. Once you experience that :bulb: moment, the basics will suddenly be very intuitive - at least it was for me.

I'd recommend on just learning the few things you need to do, and then learn more advanced stuff if you need it.

You can think of a Git repository as a structure of normal files and folders, except that it can't (not true, but that's the basic idea) be modified. Any changes will have to be written to a new "instance" of said structure instead. So, what you have are loads and loads of file- and folder structure snapshots/instances, and these are what's called commits. Each have their own SHA-1 commit id. Each time you make a new commit, you make a new "snapshot". Every commit/snapshot also knows the id of it's "parent" - the commit that came before it. Since the commit is unmodifiable, not even the pointer to the parent can be modified.

Pull requests is a GitHub concept, not a Git concept. What it really is, is a Git branch. Both Git branches and Git tags are mere pointers to a certain commit id. The rest of the "pull request" concept is just what you see here, the ability to comment, approve and otherwise collaborate around said branch. While tags can't be modified (they will always point to the exact commit id that they were created for), branches can be. That means that a branch is a pointer that can be told to point to another commit.

When you have a PR that you want to "merge" you really have one branch that you want to merge/combine with another branch (usually `master`).

When you create commits, they will always point to whatever commit is the "active commit" when you make the commit as their parent. When "merging" branches, Git will have to go back and find the most recent commit that is shared between the two branches. The list of commits you see listed on this page is simply the commits that are added to the branch in question (here `customisable_additional_directory` from repository `maciekberry`) that doesn't exist in `master`. As such, when `master` changes, so does the difference between the two, and you have to modify your branch in such a way that only the commits you want to constitute the PR are the difference.

This can be done in one of two ways: You can "merge" `master` into your branch, or you can rebase your branch on top of `master`.

Merging in Git is simply making some special kind of "magical" commits that doesn't follow the normal rules. In addition to representing a certain state of the file- and folder structure, they have multiple parents. As a consequence, you can only consider the changes for a merge commit relative to one of its parents, because the differences will be different depending on which parent you compare with. Because of this, merge commits can be very difficult to read when reading a Git history and while trying to figure out why something ended up as it did. But, merge commits offer an "easy way out" for getting two branches combined. Git simply combines the code from the two branches to the best of its ability, and if the same code is changed in both branches, you get a "merge conflict" and a human has to figure out what the combined code should be.

Rebasing is much more structured, although some find it harder to understand. What it is, is basically changing only the parent pointers for a set of existing commits. Since a commit can't be modified, copying a commit and assigning a new parent will create a whole new commit with a new id. In my opinion, rebasing is the correct approach to situations like this. What you want to keep is only the changes/commits that you have made and which you intended to be a part of the PR. Everything else you simply want to disregard. So, the only thing you really need to do is to change the parent pointer of the *first* of your commits to point to the *last* commit of the `master` branch as it is now. By doing that, your commits will follow immediately after the last commit in master, and all Git has to do to "merge" it is to add your commits to the "head" of master.

However, since changing the parent pointer of your first commit means it will become a new commit (with a new id), the next commit also has to be changed, and so it goes again and again. This means that all the commits will have to be assigned new parents and get new id's, while keeping the code changes the same.

Looking at the above commits, it seems that you started out making commits having commit c31e8cb204b80c10d3502155df7413920aee5d78 checked out (which was the "head" of `master` at some point). Right now, c56b6af4b060da15053ce2690a504c20beb524b0 is the "head" of `master`, so your first commit must point to that instead.

Rebasing will let you do this. In addition, you have some commits that's just "clutter" (correcting mistakes, reversals etc) that doesn't make much sense to put into UMS' Git history. Rebasing will let you fix that as well.

To rebase your branch to point to the latest `master` commit is very easy. You simply need to tell Git to rebase on a given branch, `master`, and Git will automatically use the current "head" of that branch as the parent of your first commit. It will also figure out the commits that are currently shared between both branches so that they won't be "duplicated".

The actual command for doing this varies on your local setup though. When you "clone" a repository to your computer, a copy of the repository you're cloning is created on your harddrive. From this point on, this is your "local repository". In addition you can define a number of "remote repositories" that has some URL on which they can be reached. This will allow you to copy commits between your local and remote repositories. By default, when you clone a repository, Git will create one remote called "origin" for you, that points to the repository you cloned. You can add others though.

In the case of GitHub PR, this can be a bit confusing. Since you don't have write permissions to UMS' GitHub repository, you first have to "fork" the repository to your GitHub account. Forking simply means copying the repository while GitHub keeps track of what repository is the origin of that repository. So, after having done this, you have 3 UMS repositories to deal with: Your local repository, UniversalMediaServer/UniversalMediaServer and maciekberry/UniversalMediaServer. You have write access to two of them. Since GitHub doesn't have any access to your local repository, the branch that is the base of a PR is a branch in maciekberry/UniversalMediaServer (maciekberry/UniversalMediaServer:customisable_additional_directory in this case). To change the PR, this is the branch you need to change. But, to change that, you first need to "push" the changed from your local repository to this repository.

Depending on at what stage you cloned UMS, your "origin" will either point to UniversalMediaServer/UniversalMediaServer or to maciekberry/UniversalMediaServer. To see where your remotes "point", simply run 
```powershell
git remote -v
```

I prefer to let "origin" point to "my GitHub fork" and to let "upstream" point to the repository I want to submit code to in such situations. Assuming this logic, doing the rebasing is as simple as:

```git checkout customisable_additional_directory``` (unless you already have that local branch checked out, ```git status``` will tell you this and more)
```git fetch upstream``` (tell Git to update your local cache of said remote repository)
```git rebase upstream/master``` (tell git to rebase your branch so that your first commit points to the last commit of `upstream/master`)

When "changing parents" in this way, conflicts that Git can't figure out might arise because the "new" parent is different that the "old", in which case the rebasing will halt. If that's the case, you'll have to figure out how to "resolve" it and then commit. Once that's done, you can continue the rebasing with 
`git rebase --continue`. Once the rebase is completed, you need to "force push" (`git push --force-with-lease`) the rebased branch to the "origin" remote. That will update this PR to reflect the changes.

To remove the "clutter commits" you can use "interactive rebasing". It will let you combine, reorder or discard commits. I can't do a full explanation of that process here, but I think you'll find many guides for how to do this. Once the interactive rebasing is done, again "force push".

It's easy to lose track or do something wrong when rebasing when you're not used to the process. Since commits are unmodifiable and branches are just pointers, it's very easy to protect yourself though. If I'm about to do something that I'm not sure will end up as I intended it, I simply make a "backup" first. If you create a new branch, the new branch will point to the currently checked-out commit. That means that before you do any rebasing, while `customisable_additional_directory` is checked out, simply run `git branch customisable_additional_directory_backup`. That will create a branch that you can always go back to if you get lost. If your rebasing has "gone bad" before you're finished, you can simply run `git rebase --abort` and you're back where you started, but if you discover that something went wrong after the rebasing has completed, you can simply run `git reset --hard customisable_additional_directory_backup` while being checked out to `customisable_additional_directory`. This will reset `customisable_additional_directory` to point to the same commit as it did when you started, *and* overwrite your "working tree" (the files and folders in your local file system) so that it is restored back to the backed-up state.

Much, much more can be said about Git, and although this has gotten a bit long already, I only intended to give you enough insight to find your way. It might have achieved the opposite, but one can always hope for the best :wink: 

There are many tools that can make working with Git easier out there, but depending on platform and preferences it's hard to say what's the best solution. On Windows I use TortoiseGit. It's not perfect, and I don't like to do rebasing using TortoiseGit (so I use the command line for that), but for most other things I find it very helpful.

**Comment by maciekberry** [comment]
Thank you @Nadahar for this message. Be assured that you have achieved the goal you wanted.

The most urgent problem is now my miserable pull request for Virtual Folders. @SubJunk has proposed to clean it up by himself (which I have accepted as the most efficient solution for the short-term). If I understand correctly you want me to clean it up myself.

**Comment by Nadahar** [comment]
@maciekberry Hehe, good. It's not up to me if you or @SubJunk should resolve it, I just wanted to give you the ability to handle it yourself. 

If doubt @SubJunk will mind of you do it though, as that means he won't have to :wink:

**Comment by maciekberry** [comment]
_Roma locuta causa finita est!_ 

Thanks @valib

**Comment by SubJunk** [comment]
@valib thanks, by the way do you have the ability to merge? Please feel free to merge things after approving them

**Pull request merged**